BundleTransport/app/src/main/java/com/ddd/bundletransport/GrpcReceiveTask.java

```diff
@@ -21,47 +21,45 @@
 import io.grpc.ManagedChannelBuilder;
 import io.grpc.stub.StreamObserver;
 
-public class GrpcReceiveTask implements Runnable {
+public class GrpcReceiveTask {
     private final static String TAG = "dddTransport";
-    private Context context;
     private String host, receiveDir, transportId;
     private int port;
     private boolean receiveBundles, statusComplete;
     private ManagedChannel channel;
-    private Function<Exception, Void> callback;
 
-    public GrpcReceiveTask(Context context, String host, String port, String transportId, Function<Exception, Void> callback) {
+    public GrpcReceiveTask(String host, int port, String transportId, String receiveDir) {
         Log.d(TAG, "initializing grpcreceivetask...");
-        this.context = context;
         this.host = host;
-        this.port = Integer.parseInt(port);
+        this.port = port;
         this.transportId = transportId;
-        this.receiveDir = this.context.getExternalFilesDir(null) + "/BundleTransmission/client";
-        this.callback = callback;
+        this.receiveDir = receiveDir;
     }
 
-    @Override
-    public void run() {
+    public Exception run() {
         Exception thrown = null;
         try {
             executeTask();
+            Log.d(TAG, "Executed receive task");
         } catch (Exception e) {
             thrown = e;
         }
-        callback.apply(thrown);
+
         try {
             postExecuteTask();
         } catch (InterruptedException e) {
             Log.e(TAG, "Failed to shutdown GrpcReceiveTask channel: " + e.getMessage());
         }
+
+        return thrown;
     }
 
-    private void executeTask() {
+    private void executeTask() throws Exception {
         Log.d(TAG, "executing grpcreceivetask...");
 
         channel = ManagedChannelBuilder.forAddress(host, port).usePlaintext().build();
         BundleServiceGrpc.BundleServiceStub stub = BundleServiceGrpc.newStub(channel);
-
+        Throwable[] thrown = new Throwable[1];
```

Comment:
it might be good to put a comment that we are using an array because it is set in the anonymous class below.